waiter/src/waiter/scheduler.clj

```diff
@@ -642,3 +646,191 @@
+            {started-instance-ids :started starting-instances-ids :starting}
+            (group-by #(if (healthy-instance-id? %) :started :starting)
+                      (keys starting-instance-id->start-timestamp'))
+            starting-instance-id->start-timestamp''
+            (select-keys starting-instance-id->start-timestamp' starting-instances-ids)]
+        ;; Warn about any new instances for which we didn't track the scheduling time
+        (when (pos? unmatched-instance-count)
+          (log/warn unmatched-instance-count
+                    "untracked scheduled instances discovered in instance launch metric loop for service"
+                    service-id))
+        ;; Report schedule time for now-known instances
+        (doseq [start-time matched-start-times]
+          (let [duration (metrics/duration-between start-time current-time)]
+            (metrics/report-duration waiter-schedule-timer duration)
+            (metrics/report-duration service-schedule-timer duration)))
+        ;; Report startup time for instances that are now healthy
+        (doseq [instance-id started-instance-ids]
+          (let [start-time (starting-instance-id->start-timestamp' instance-id)]
+            (metrics/report-duration service-startup-timer start-time current-time)))
+        ;; tracker-state'
+        (assoc tracker-state
+               :instance-counts instance-counts'
+               :instance-scheduling-start-times (vec instance-scheduling-start-times'')
+               :known-instance-ids known-instance-ids'
+               :starting-instance-id->start-timestamp starting-instance-id->start-timestamp'')))))
+
+(defn build-initial-service-launch-trackers
+  "Build the initially-observed state for the launch-metrics-maintainer's state loop
+   from the first state observed on the router-state-updates channel."
+  [{:keys [iteration service-id->healthy-instances service-id->instance-counts
+           service-id->unhealthy-instances] :as initial-router-state}]
+  (let [initial-service-ids (-> service-id->instance-counts keys set)
+        initial-service-id->launch-tracker
+        (pc/for-map [service-id initial-service-ids]
+          service-id
+          (let [instance-counts (get service-id->instance-counts service-id)
+                known-instance-ids
+                (set (for [service-id->instances [service-id->healthy-instances
+                                                  service-id->unhealthy-instances]
+                           instance (get service-id->instances service-id)]
+                       (:id instance)))]
+            (make-launch-tracker service-id instance-counts known-instance-ids)))]
+    (log/info "Starting launch metrics loop on iteration" iteration
+              "with services:" initial-service-ids)
+    {:known-service-ids initial-service-ids
+     :previous-iteration iteration
+     :service-id->launch-tracker initial-service-id->launch-tracker}))
+
+(def no-previous-iteration-value -1)
```

Comment:
Nitpick: Could we remove `no-previous-iteration-value` and use `nil` where `no-previous-iteration-value` is used?  `nil` seems more appropriate than `-1`.

Reply:
Done!